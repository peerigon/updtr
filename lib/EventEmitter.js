"use strict";

var events = require("events");

var EventEmitter = events;

if (typeof EventEmitter !== "function") {
    EventEmitter = events.EventEmitter;
}

module.exports = EventEmitter;
