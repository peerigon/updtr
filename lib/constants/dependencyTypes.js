"use strict";

const REGULAR = 0;
const DEV = 1;
const OPTIONAL = 2;

function fromNpm(type) {
    switch (type) {
        case "dependencies":
            return REGULAR;
        case "devDependencies":
            return DEV;
        case "optionalDependencies":
            return OPTIONAL;
        default:
            throw new Error(`Unsupported dependency type: ${ type }`);
    }
}

exports.REGULAR = REGULAR;
exports.DEV = DEV;
exports.OPTIONAL = OPTIONAL;

exports.fromNpm = fromNpm;
// yarn uses the same strings. Change this if necessary.
exports.fromYarn = fromNpm;
