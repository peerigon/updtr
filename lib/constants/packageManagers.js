"use strict";

const NPM = "npm";
const YARN = "yarn";
const SUPPORTED = [NPM, YARN];

exports.NPM = NPM;
exports.YARN = YARN;
exports.SUPPORTED = SUPPORTED;
