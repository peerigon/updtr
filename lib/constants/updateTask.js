"use strict";

const WANTED = "wanted";
const LATEST = "latest";

exports.WANTED = WANTED;
exports.LATEST = LATEST;
