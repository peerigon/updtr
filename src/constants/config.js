// The first value in arrays is the default value

export const USE_NPM = "npm";
export const USE_YARN = "yarn";
export const USE_OPTIONS = [USE_NPM, USE_YARN];
export const UPDATE_TO_LATEST = "latest";
export const UPDATE_TO_NON_BREAKING = "non-breaking";
export const UPDATE_TO_WANTED = "wanted";
export const UPDATE_TO_OPTIONS = [
    UPDATE_TO_LATEST,
    UPDATE_TO_NON_BREAKING,
    UPDATE_TO_WANTED,
];
export const SAVE_SMART = "smart";
export const SAVE_CARET = "caret";
export const SAVE_EXACT = "exact";
export const SAVE_OPTIONS = [SAVE_SMART, SAVE_CARET, SAVE_EXACT];
