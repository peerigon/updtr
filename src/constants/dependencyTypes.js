export const REGULAR = "dependencies";
export const DEV = "devDependencies";
export const OPTIONAL = "optionalDependencies";

export function fromNpm(type) {
    switch (type) {
        case "dependencies":
            return REGULAR;
        case "devDependencies":
            return DEV;
        case "optionalDependencies":
            return OPTIONAL;
        default:
            throw new Error(`Unsupported dependency type: ${ type }`);
    }
}

export {
    // yarn uses the same strings. Change this if necessary.
    fromNpm as fromYarn,
};
