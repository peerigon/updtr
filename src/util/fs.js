import fs from "fs";
import pify from "pify";

export default pify(fs);
