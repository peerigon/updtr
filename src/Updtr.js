import EventEmitter from "events";
import path from "path";
import fs from "./util/fs";
import {
    USE_OPTIONS,
    UPDATE_TO_OPTIONS,
    SAVE_OPTIONS,
} from "./constants/config";
import exec from "./exec/exec";
import cmds from "./exec/cmds";
import parse from "./exec/parse";
import {
    RequiredOptionMissingError,
    OptionValueNotSupportedError,
    YarnWithCustomRegistryError,
} from "./errors";

// node v4 has no dedicated constants object.
// Remove this if node v4 is not supported anymore.
const FS_CONSTANTS = fs.constants === undefined ? fs : fs.constants;

function checkCwd(cwd) {
    if (typeof cwd !== "string") {
        throw new RequiredOptionMissingError("cwd", cwd);
    }
}

function checkUse(use) {
    if (USE_OPTIONS.indexOf(use) === -1) {
        throw new OptionValueNotSupportedError("use", use);
    }
}

function checkUpdateTo(updateTo) {
    if (UPDATE_TO_OPTIONS.indexOf(updateTo) === -1) {
        throw new OptionValueNotSupportedError("updateTo", updateTo);
    }
}

function checkSave(save) {
    if (SAVE_OPTIONS.indexOf(save) === -1) {
        throw new OptionValueNotSupportedError("save", save);
    }
}

function checkForYarnWithCustomReg(packageManager, registry) {
    if (packageManager === "yarn" && registry !== undefined) {
        throw new YarnWithCustomRegistryError();
    }
}

export default class Updtr extends EventEmitter {
    /**
     * The config passed-in here should look identically to the CLI config.
     * Dash-cased properties should be renamed to camelCased.
     * The goal is to replicate the API of the CLI as close as possible so users don't
     * have to guess the options.
     *
     * @param {UpdtrConfig} config
     */
    constructor(config) {
        super();

        const cwd = config.cwd;
        const registry = config.registry;
        const packageManager = config.use === undefined ?
            USE_OPTIONS[0] :
            config.use;
        const updateTo = config.updateTo === undefined ?
            UPDATE_TO_OPTIONS[0] :
            config.updateTo;
        const exclude = Array.isArray(config.exclude) === true ?
            config.exclude :
            [];
        const save = config.save === undefined ? SAVE_OPTIONS[0] : config.save;

        checkCwd(cwd);
        checkUse(packageManager);
        checkForYarnWithCustomReg(packageManager, registry);
        checkUpdateTo(updateTo);
        checkSave(save);

        this.config = {
            cwd,
            use: packageManager,
            exclude,
            test: config.test,
            registry,
            updateTo,
            save,
        };
        this.cmds = cmds[packageManager];
        this.parse = parse[packageManager];

        if (typeof config.test === "string") {
            this.cmds = {
                ...this.cmds,
                test: () => config.test,
            };
        }
    }
    async canAccessPackageJson() {
        let result = true;

        try {
            await fs.access(
                path.join(this.config.cwd, "package.json"),
                FS_CONSTANTS.R_OK | FS_CONSTANTS.W_OK // eslint-disable-line no-bitwise
            );
        } catch (err) {
            result = false;
        }

        return result;
    }
    exec(cmd) {
        return exec(this.config.cwd, cmd);
    }
    readFile(filenameInCwd) {
        return fs.readFile(path.join(this.config.cwd, filenameInCwd), "utf8");
    }
    writeFile(filenameInCwd, contents) {
        return fs.writeFile(
            path.join(this.config.cwd, filenameInCwd),
            contents
        );
    }
    dispose() {
        this.removeAllListeners();
    }
}
