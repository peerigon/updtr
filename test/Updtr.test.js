import EventEmitter from "events";
import path from "path";
import Updtr from "../src/Updtr";
import fs from "../src/util/fs";
import {
    RequiredOptionMissingError,
    OptionValueNotSupportedError,
    YarnWithCustomRegistryError,
} from "../src/errors";
import { USE_YARN, UPDATE_TO_OPTIONS } from "../src/constants/config";
import temp from "./helpers/temp";
import FakeUpdtr from "./helpers/FakeUpdtr";

describe("new Updtr()", () => {
    it("should return an updtr with expected shape", () => {
        const updtr = new Updtr(FakeUpdtr.baseConfig);

        expect(updtr).toBeInstanceOf(EventEmitter);
        // Check for properties that are not tested in this unit test
        expect(updtr).toHaveProperty("cmds");
        expect(updtr).toHaveProperty("parse");
    });
    describe(".config", () => {
        it("should match the default shape", () => {
            const updtr = new Updtr(FakeUpdtr.baseConfig);

            expect(updtr.config).toMatchSnapshot();
        });
        describe(".updateTo", () => {
            UPDATE_TO_OPTIONS.forEach(updateTo => {
                it(`should be ${ updateTo } if given`, () => {
                    const config = { ...FakeUpdtr.baseConfig, updateTo };
                    const updtr = new Updtr(config);

                    expect(updtr.config).toHaveProperty("updateTo", updateTo);
                });
            });
        });
        describe(".exclude", () => {
            it("should match the given exclude array", () => {
                const config = {
                    ...FakeUpdtr.baseConfig,
                    exclude: ["a", "b", "c"],
                };
                const updtr = new Updtr(config);

                expect(updtr.config).toHaveProperty("exclude", ["a", "b", "c"]);
            });
        });
        describe(".registry", () => {
            it("should be set if a custom registry was given", () => {
                const config = {
                    ...FakeUpdtr.baseConfig,
                    registry: "http://example.com",
                };
                const updtr = new Updtr(config);

                expect(updtr.config).toHaveProperty(
                    "registry",
                    "http://example.com"
                );
            });
        });
        describe(".use", () => {
            it("should be 'yarn' if specified", () => {
                const config = { ...FakeUpdtr.baseConfig, use: USE_YARN };
                const updtr = new Updtr(config);

                expect(updtr.config).toHaveProperty("use", USE_YARN);
            });
        });
    });
    describe(".cmds", () => {
        describe(".test()", () => {
            describe("when a custom test command was given", () => {
                it("should return the custom test command", () => {
                    const test = "custom test command";
                    const config = { ...FakeUpdtr.baseConfig, test };
                    const updtr = new Updtr(config);

                    expect(updtr.cmds.test()).toBe(test);
                });
            });
        });
    });
    describe(".canAccessPackageJson()", () => {
        describe("when there is no package.json in the cwd", () => {
            it("should return false", async () => {
                const cwd = __dirname;
                const updtr = new Updtr({ ...FakeUpdtr.baseConfig, cwd });

                expect(await updtr.canAccessPackageJson()).toBe(false);
            });
        });
        describe("when there is a package.json in the cwd", () => {
            it("should return false", async () => {
                const cwd = path.join(__dirname, "fixtures", "empty");
                const updtr = new Updtr({ ...FakeUpdtr.baseConfig, cwd });

                expect(await updtr.canAccessPackageJson()).toBe(true);
            });
        });
    });
    describe(".exec()", () => {
        it("should exec the command in the given cwd", async () => {
            const cwd = __dirname;
            const updtr = new Updtr({ ...FakeUpdtr.baseConfig, cwd });
            const cmd = 'node -e "console.log(process.cwd())"';
            const result = await updtr.exec(cmd);

            expect(result.stdout).toBe(cwd + "\n"); // also just \n on windows
        });
    });
    describe("readFile()", () => {
        it("should read a file as utf8 string relative to cwd", async () => {
            const testContent = "This is a test";
            const testJs = "test.js";
            const cwd = await temp.mkdir("updtr-readFile-1");
            const updtr = new Updtr({
                cwd,
            });

            await fs.writeFile(path.join(cwd, testJs), testContent);
            expect(await updtr.readFile(testJs)).toBe(testContent);
        });
        it("should not swallow errors", async () => {
            const testJs = "test.js";
            const cwd = await temp.mkdir("updtr-readFile-2");
            const updtr = new Updtr({
                cwd,
            });
            let givenErr;

            try {
                await updtr.readFile(testJs);
            } catch (err) {
                givenErr = err;
            }

            expect(givenErr).toHaveProperty("message");
            expect(givenErr.message).toMatch(/no such file or directory/);
        });
    });
    describe("writeFile()", () => {
        it("should write a file as utf8 string relative to cwd", async () => {
            const testContent = "This is a test";
            const testJs = "test.js";
            const cwd = await temp.mkdir("updtr-writeFile-1");
            const updtr = new Updtr({
                cwd,
            });

            await updtr.writeFile(testJs, testContent);

            expect(await fs.readFile(path.join(cwd, testJs), "utf8")).toBe(
                testContent
            );
        });
        it("should not swallow errors", async () => {
            const cwd = await temp.mkdir("updtr-writeFile-2");
            const updtr = new Updtr({
                cwd,
            });
            let givenErr;

            try {
                await updtr.writeFile("");
            } catch (err) {
                givenErr = err;
            }

            expect(givenErr).toHaveProperty("message");
            expect(givenErr.message).toMatch(
                /illegal operation on a directory/
            );
        });
    });
    describe(".dispose()", () => {
        it("should remove all event listeners", () => {
            const updtr = new Updtr(FakeUpdtr.baseConfig);

            updtr.on("test", () => {
                throw new Error("Should not be called");
            });
            updtr.dispose();
            updtr.emit("test");
        });
    });
    describe("errors", () => {
        it("should throw if a cwd is missing", () => {
            const config = { ...FakeUpdtr.baseConfig };

            delete config.cwd;

            expect(() => new Updtr(config)).toThrow(RequiredOptionMissingError);
        });
        it("should throw if the use option is not supported", () => {
            const config = { ...FakeUpdtr.baseConfig, use: "bower" };

            expect(() => new Updtr(config)).toThrow(
                OptionValueNotSupportedError
            );
        });
        it("should throw if the updateTo option is unknown", () => {
            const config = {
                ...FakeUpdtr.baseConfig,
                updateTo: "something-else",
            };

            expect(() => new Updtr(config)).toThrow(
                OptionValueNotSupportedError
            );
        });
        it("should throw if the save option is unknown", () => {
            const config = { ...FakeUpdtr.baseConfig, save: "something-else" };

            expect(() => new Updtr(config)).toThrow(
                OptionValueNotSupportedError
            );
        });
        it("should throw if package manager is yarn and there is a custom registry set", () => {
            const config = {
                ...FakeUpdtr.baseConfig,
                registry: "http://example.com",
                use: USE_YARN,
            };

            expect(() => new Updtr(config)).toThrow(
                YarnWithCustomRegistryError
            );
        });
    });
});

afterAll(temp.cleanup);
