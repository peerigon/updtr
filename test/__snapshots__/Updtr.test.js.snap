// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new Updtr() .config should match the default shape 1`] = `
Object {
  "cwd": "/updtr/test/cwd",
  "exclude": Array [],
  "registry": undefined,
  "save": "smart",
  "test": undefined,
  "updateTo": "latest",
  "use": "npm",
}
`;
